torch/_dynamo/variables/tensor.py

```diff
@@ -681,13 +687,61 @@ def redistribute_fn_with_prim_types(x):
                 mutable_local=variables.base.MutableLocal(),
                 **options,
             )
+
             if not self.source:
                 # Intermediary
-                unimplemented("Intermediary tensors with registered hooks - NYI")
-            else:
-                assert (
-                    fn_var.source
-                ), "Unreachable - See unimplemented for lambdas above"
+                src = fn_var.source
+                if (
+                    not src
+                    and isinstance(fn_var, variables.functions.FunctoolsPartialVariable)
+                    and fn_var.func.source
+                ):
+                    src = fn_var.func.source
+
+                if not src:
+                    unimplemented("No source for register_hook target fn")
+
+                tx.output.guards.add(src.make_guard(GuardBuilder.ID_MATCH))
```

Comment:
This still doesn't seem like enough checking. For a functools partial you need to also ensure that the args are guarded on, because you're burning in this particular functools.partial application into the graph; if they vary from run to run you will be stale.

Reply:
We do propagate the guards from the fn! See the lines above:

```
            # Guards from the fn_var
            options.update(VariableTracker.propagate(fn_var))
```